src/Config/Converters/DataSourceConverterFactory.cs

```diff
@@ -0,0 +1,123 @@
+using Azure.DataApiBuilder.Config.ObjectModel;
+
+namespace Azure.DataApiBuilder.Config.Converters;
+
+internal class DataSourceConverterFactory : JsonConverterFactory
+{
+    // Determines whether to replace environment variable with its
+    // value or not while deserializing.
+    private bool _replaceEnvVar;
+
+    /// <inheritdoc/>
+    public override bool CanConvert(Type typeToConvert)
+    {
+        return typeToConvert.IsAssignableTo(typeof(DataSource));
+    }
+
+    /// <inheritdoc/>
+    public override JsonConverter? CreateConverter(Type typeToConvert, JsonSerializerOptions options)
+    {
+        return new DataSourceConverter(_replaceEnvVar);
+    }
+
+    /// <param name="replaceEnvVar">Whether to replace environment variable with its
+    /// value or not while deserializing.</param>
+    internal DataSourceConverterFactory(bool replaceEnvVar)
+    {
+        _replaceEnvVar = replaceEnvVar;
+    }
+
+    private class DataSourceConverter : JsonConverter<DataSource>
+    {
+        // Determines whether to replace environment variable with its
+        // value or not while deserializing.
+        private bool _replaceEnvVar;
+
+        /// <param name="replaceEnvVar">Whether to replace environment variable with its
+        /// value or not while deserializing.</param>
+        public DataSourceConverter(bool replaceEnvVar)
+        {
+            _replaceEnvVar = replaceEnvVar;
+        }
+
+        public override DataSource? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            DataSource dataSource = new(DatabaseType.MSSQL, string.Empty, null);
+            if (reader.TokenType is JsonTokenType.StartObject)
+            {
+
+                while (reader.Read())
+                {
+                    if (reader.TokenType is JsonTokenType.EndObject)
+                    {
+                        break;
+                    }
+
+                    if (reader.TokenType is JsonTokenType.PropertyName)
+                    {
+                        string propertyName = reader.GetString() ?? string.Empty;
+                        reader.Read();
+                        switch (propertyName)
+                        {
+                            case "database-type":
+                                dataSource = dataSource with { DatabaseType = EnumExtensions.Deserialize<DatabaseType>(reader.DeserializeString(_replaceEnvVar)!) };
+                                break;
+                            case "connection-string":
+                                dataSource = dataSource with { ConnectionString = reader.DeserializeString(replaceEnvVar: _replaceEnvVar)! };
+                                break;
+                            case "options":
+                                Dictionary<string, JsonElement> optionsDict = new();
+                                while (reader.Read() && reader.TokenType is not JsonTokenType.EndObject)
+                                {
```

Comment:
Can't the outer while have a similar condition `reader.Read() && reader.TokenType is not JsonTokenType.EndObject` instead of adding a `break` as the 1st statement inside the loop?

Reply:
yep, thanks for pointing